### PR TITLE
feat: `avm/ptn/authorization/role-assignment`

### LINKED ISSUE #775: [Module Proposal]: `avm/ptn/authorization/role-assignment`
### Check for previous/existing GitHub issues/module proposals

- [X] I have checked for previous/existing GitHub issues/module proposals.

### Check this module doesn't already exist in the module indexes

- [X] I have checked for that this module doesn't already exist in the module indexes; or I'm proposing the module to be migrated from CARML/TFVM.

### Bicep or Terraform?

Bicep

### Module Classification?

Pattern Module

### Module Name

avm/ptn/authorization/role-assignment

### Module Details

Module migration from CARML to AVM

### Do you want to be the owner of this module?

Yes

### Module Owner's GitHub Username (handle)

arnoldna

### (Optional) Secondary Module Owner's GitHub Username (handle)

_No response_

### PR DESCRIPTION
**Pull request by arnoldna**
## Description

Migrating module from CARML to AVM

Closes https://github.com/Azure/Azure-Verified-Modules/issues/775

## Pipeline Reference

[avm.ptn.authorization.role-assignment](https://github.com/arnoldna/bicep-registry-modules/actions/workflows/avm.ptn.authorization.role-assignment.yml)
## Type of Change

- [ ] Update to CI Environment or utlities (Non-module effecting changes)
- [X] Azure Verified Module updates:
  - [ ] Bugfix containing backwards compatible bug fixes, and I have NOT bumped the MAJOR or MINOR version in `version.json`:
    - [ ] Someone has opened a bug report issue, and I have included "Closes #{bug_report_issue_number}" in the PR description.
    - [ ] The bug was found by the module author, and no one has opened an issue to report it yet.
  - [ ] Feature update backwards compatible feature updates, and I have bumped the MINOR version in `version.json`.
  - [ ] Breaking changes and I have bumped the MAJOR version in `version.json`.
  - [X] Update to documentation

## Checklist

- [X] I'm sure there are no other open Pull Requests for the same update/change
- [X] I have run `Set-AVMModule` locally to generate the supporting module files.
- [ ] My corresponding pipelines / checks run clean and green without any errors or warnings

### CODE REVIEW

File: .github/workflows/avm.ptn.authorization.role-assignment.yml [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+      - ".github/actions/templates/avm-**"
+      - ".github/workflows/avm.template.module.yml"
+      - ".github/workflows/avm.ptn.authorization.role-assignment.yml"
+      - "avm/res/analysis-services/server/**"
+      - "avm/utilities/pipelines/**"
+      - "!avm/utilities/pipelines/platform/**"
+      - "!*/**/README.md"
+
+env:
+  modulePath: "avm/res/analysis-services/server"
```
Comment (eriqua) [inline .github/workflows/avm.ptn.authorization.role-assignment.yml:36]: Please make sure all variables refer to the new module

File: .github/ISSUE_TEMPLATE/avm_module_issue.yml [hunk 2]
```diff
@@ -44,6 +44,7 @@ body:
         # - "avm/ptn/avd-lza/session-hosts"
         # - "avm/ptn/security/security-center"
         # - "avm/res/aad/domain-service"
+        - "avm/ptn/authorization/role-assignments"
```
Comment (eriqua) [inline .github/ISSUE_TEMPLATE/avm_module_issue.yml:47]: This should be placed in alphabetical order. Static validation should fail on this. Isn't that the case?

File: .github/workflows/avm.res.sql.managed-instance.yml
Comment (eriqua) [inline .github/workflows/avm.res.sql.managed-instance.yml:1]: This sneaked in, please only include files related to this module

**Review by eriqua — COMMENTED** [review]
@arnoldna could you please attach a pipeline badge?

File: avm/ptn/authorization/role-assignment/management-group/version.json
Comment (AlexanderSehr) [inline avm/ptn/authorization/role-assignment/management-group/version.json:1]: Child modules should not have `version.json` files

File: avm/ptn/authorization/role-assignment/management-group/main.bicep
Comment (AlexanderSehr) [inline avm/ptn/authorization/role-assignment/management-group/main.bicep:1]: As this is 'technically' not an actual child module (in the sence that it's a child-resource), I recommend to
- Rename all 3 module files from `main.bicep` to `management-group.bicep` , `subscription.bicep` & `resource-group.bicep`   
- Move all 3 files in a `/modules/` subfolder in the module's root
- Delete the remaining files (namely the readme.md files + the version.json files)

**Review by AlexanderSehr — CHANGES_REQUESTED** [review]

File: avm/ptn/authorization/role-assignment/resource-group/version.json
Comment (AlexanderSehr) [inline avm/ptn/authorization/role-assignment/resource-group/version.json:1]: Child modules should not have `version.json` files

File: avm/ptn/authorization/role-assignment/subscription/version.json
Comment (AlexanderSehr) [inline avm/ptn/authorization/role-assignment/subscription/version.json:1]: Child modules should not have `version.json` files

File: avm/ptn/authorization/role-assignment/tests/e2e/rg.common/main.test.bicep [hunk 3]
```diff
@@ -1,4 +1,4 @@
-targetScope = 'subscription'
+targetScope = 'managementGroup'
 metadata name = 'Role Assignments (Resource Group)'
```
Comment (AlexanderSehr) [inline avm/ptn/authorization/role-assignment/tests/e2e/rg.common/main.test.bicep:2]:
```suggestion
metadata name = 'Role Assignments (Resource Group scope)'
```
to align with other test cases

File: avm/ptn/authorization/role-assignment/tests/e2e/rg.min/main.test.bicep [hunk 4]
```diff
@@ -1,4 +1,4 @@
-targetScope = 'subscription'
+targetScope = 'managementGroup'
 metadata name = 'Role Assignments (Resource Group)'
```
Comment (AlexanderSehr) [inline avm/ptn/authorization/role-assignment/tests/e2e/rg.min/main.test.bicep:2]:
```suggestion
metadata name = 'Role Assignments (Resource Group scope)'
```

File: avm/ptn/authorization/role-assignment/tests/e2e/sub.min/main.test.bicep [hunk 5]
```diff
@@ -46,9 +57,10 @@ module testDeployment '../../../main.bicep' = {
   name: '${uniqueString(deployment().name)}-test-${serviceShort}'
   params: {
     principalId: nestedDependencies.outputs.managedIdentityPrincipalId
-    roleDefinitionIdOrName: 'Resource Policy Contributor'
+    roleDefinitionIdOrName: '/providers/Microsoft.Authorization/roleDefinitions/c2f4ef07-c644-48eb-af81-4b1b4947fb11'
```
Comment (AlexanderSehr) [inline avm/ptn/authorization/role-assignment/tests/e2e/sub.min/main.test.bicep:60]:
```suggestion
    roleDefinitionIdOrName: subscriptionResourceId('Microsoft.Authorization/roleDefinitions', 'acdd72a7-3385-48ef-bd42-f606fba81ae7')
```

a bit easier to use

File: avm/ptn/authorization/role-assignment/tests/e2e/sub.min/main.test.bicep [hunk 6]
```diff
@@ -46,9 +57,10 @@ module testDeployment '../../../main.bicep' = {
...
     principalType: 'ServicePrincipal'
     location: resourceLocation
     subscriptionId: subscriptionId
+    resourceGroupName: resourceGroupName
```
Comment (AlexanderSehr) [inline avm/ptn/authorization/role-assignment/tests/e2e/sub.min/main.test.bicep:64]: Why the resourceGroupName? Does it then not apply the roleAssignment to the RG scope, as opposed to the subscription scope that we want to test?

File: avm/ptn/authorization/role-assignment/tests/e2e/sub.common/main.test.bicep [hunk 7]
```diff
@@ -62,5 +62,6 @@ module testDeployment '../../../main.bicep' = {
     principalType: 'ServicePrincipal'
     location: resourceLocation
     subscriptionId: subscriptionId
+    resourceGroupName: resourceGroupName
```
Comment (AlexanderSehr) [inline avm/ptn/authorization/role-assignment/tests/e2e/sub.common/main.test.bicep:65]: Why the resourceGroupName? Does it then not apply the roleAssignment to the RG scope, as opposed to the subscription scope that we want to test?

**Review by AlexanderSehr — CHANGES_REQUESTED** [review]

**Review by AlexanderSehr — DISMISSED** [review]

File: .github/workflows/avm.ptn.authorization.role-assignment.yml (hunk 1 above)
Comment (AlexanderSehr) [reply .github/workflows/avm.ptn.authorization.role-assignment.yml]: Checked and fixed

File: .github/ISSUE_TEMPLATE/avm_module_issue.yml (hunk 2 above)
Comment (AlexanderSehr) [reply .github/ISSUE_TEMPLATE/avm_module_issue.yml]: Checked and fixed

File: .github/workflows/avm.res.sql.managed-instance.yml
Comment (AlexanderSehr) [reply .github/workflows/avm.res.sql.managed-instance.yml:1]: Checked and fixed

File: .github/CODEOWNERS [hunk 8]
```diff
@@ -3,6 +3,7 @@
 /scripts/ @Azure/bicep-admins @Azure/avm-core-team-technical-bicep
 /avm/ @Azure/avm-core-team-technical-bicep
 /avm/utilities/ @Azure/avm-core-team-technical-bicep
+/avm/ptn/authorization/role-assignment/ @Azure/avm-ptn-authorization-roleassignment-module-owners-bicep @Azure/avm-core-team-technical-bicep
```
[the remark this answers is not in the document]
Comment (arnoldna) [reply .github/CODEOWNERS:6]: Team has been created.

File: avm/ptn/authorization/role-assignment/tests/e2e/rg.common/dependencies.bicep
Comment (AlexanderSehr) [inline avm/ptn/authorization/role-assignment/tests/e2e/rg.common/dependencies.bicep:1]: Please rename `common` in the folder name to `max`

File: avm/ptn/authorization/role-assignment/tests/e2e/rg.common/main.test.bicep [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+
+@description('Optional. The name of the resource group to deploy for testing purposes.')
+@maxLength(90)
+param resourceGroupName string = 'dep-${namePrefix}-authorization.roleassignments-${serviceShort}-rg'
+
+@description('Optional. The location to deploy resources to.')
+param resourceLocation string = deployment().location
+
+@description('Optional. A short identifier for the kind of deployment. Should be kept short to not run into resource-name length-constraints.')
+param serviceShort string = 'arargcom'
```
Comment (AlexanderSehr) [inline avm/ptn/authorization/role-assignment/tests/e2e/rg.common/main.test.bicep:17]:
```suggestion
param serviceShort string = 'arargmax'
```

File: avm/ptn/authorization/role-assignment/tests/e2e/mg.min/main.test.bicep
Comment (AlexanderSehr) [inline avm/ptn/authorization/role-assignment/tests/e2e/mg.min/main.test.bicep:1]: Please rename `min` in the folder name to `defaults`

File: avm/ptn/authorization/role-assignment/tests/e2e/mg.common/main.test.bicep [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+
+@description('Optional. The name of the resource group to deploy for testing purposes.')
+@maxLength(90)
+param resourceGroupName string = 'dep-${namePrefix}-authorization.roleassignments-${serviceShort}-rg'
+
+@description('Optional. The location to deploy resources to.')
+param resourceLocation string = deployment().location
+
+@description('Optional. A short identifier for the kind of deployment. Should be kept short to not run into resource-name length-constraints.')
+param serviceShort string = 'aramgcom'
```
Comment (AlexanderSehr) [inline avm/ptn/authorization/role-assignment/tests/e2e/mg.common/main.test.bicep:17]:
```suggestion
param serviceShort string = 'aramgmax'
```

File: avm/ptn/authorization/role-assignment/tests/e2e/sub.common/main.test.bicep [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+
+@description('Optional. The name of the resource group to deploy for testing purposes.')
+@maxLength(90)
+param resourceGroupName string = 'dep-${namePrefix}-authorization.roleassignments-${serviceShort}-rg'
+
+@description('Optional. The location to deploy resources to.')
+param resourceLocation string = deployment().location
+
+@description('Optional. A short identifier for the kind of deployment. Should be kept short to not run into resource-name length-constraints.')
+param serviceShort string = 'arasubcom'
```
Comment (AlexanderSehr) [inline avm/ptn/authorization/role-assignment/tests/e2e/sub.common/main.test.bicep:17]:
```suggestion
param serviceShort string = 'arasubmax'
```

File: avm/ptn/authorization/role-assignment/tests/e2e/mg.min/main.test.bicep
Comment (AlexanderSehr) [reply avm/ptn/authorization/role-assignment/tests/e2e/mg.min/main.test.bicep:1]: Same for the other `defaults` tests

File: avm/ptn/authorization/role-assignment/tests/e2e/mg.common/main.test.bicep
Comment (AlexanderSehr) [inline avm/ptn/authorization/role-assignment/tests/e2e/mg.common/main.test.bicep:1]: Please rename `common` in the folder name to `max`.
Same for the other `max` tests

File: .github/workflows/avm.ptn.authorization.role-assignment.yml
Comment (AlexanderSehr) [inline .github/workflows/avm.ptn.authorization.role-assignment.yml:1]: Hey @arnoldna looks like we're pretty much done. However, somehow the workflow got deleted in main and now I can't see the workflow anymore :D Was it green last time you checked?

Comment (arnoldna) [reply .github/workflows/avm.ptn.authorization.role-assignment.yml:1]: I don't know if it was ever in BRM main but only in my local fork's main. Here is the status badge for today. [avm.ptn.authorization.role-assignment](https://github.com/arnoldna/bicep-registry-modules/actions/workflows/avm.ptn.authorization.role-assignment.yml)

**Review by AlexanderSehr — APPROVED** [review]

**Review by AlexanderSehr — APPROVED** [review]

**Pull request merged**